[AL-582] Adds clone-content command.

dustinleblanc opened the pull request:
TODO:
- Confirm the real output desired (I made something fancy that I liked!)
- Add lots more unit tests
- Add interactivity
- Probably review the AC's to see if we can flesh them out further

dustinleblanc committed:
Adds clone-content command.

dustinleblanc commented:
@TeslaDethray I think this might be ready for code review. @ari-gold we probably want to add interaction to this too?

SaraMcCutcheon commented on src/Commands/CloneCommand.php:
```diff
@@ -0,0 +1,217 @@
+    private function validateSites($origin, $target)
+    {
+        if (explode('.', $origin)[0] != explode('.', $target)[0]) {
+            throw new \Exception("Clone operations are only available within the same site.", 1);
```
If we do not permit clone operations between sites, the second parameter should not be a site/env, but just the name of an env.

dustinleblanc replied:
It'd certainly make for less code and parsing, but would `terminus command site.env env` be a canonical way to call commands? I went with the syntax of `origin.env destination.env` purely to maintain consistency. I guess having it be `terminus env:clone-content live dev --site=foo-site` might be better? The original AC had us using options for everything, which didn't feel like the way were going (though I could go back to that...)

SaraMcCutcheon commented on src/Commands/CloneCommand.php:
```diff
@@ -0,0 +1,217 @@
+    private function formatOpMethod($operation = '')
+    {
+        if (!array_key_exists($operation, $this->methods)) {
+            throw new \Exception("{$operation} does not have an available clone method.", 1);
```
We are not validating inputs in this phase. Using this method dictionary makes the processes being run less clear.

dustinleblanc replied:
I'll remove the abstraction

SaraMcCutcheon commented on src/Commands/Env/CloneContentCommand.php:
```diff
@@ -0,0 +1,26 @@
+     */
+    public function cloneContent($origin, $target)
+    {
+        $this->invokeClone($origin, $target, ['database', 'files']);
```
The logic for this function should not exist in its parent.

dustinleblanc replied:
Can you elaborate? This was done this way so that we can later do commands like `db:clone` which will wrap the the higher level command and pass in only the databse. This was based on the discussion around usage I brought up with @ari-gold. I'm curious why this would be a bad pattern.

bensheldon replied:
To also chime in, I would like to avoid abstracting command logic at this time. I'd rather see duplicate code and a better understanding of the new command system once we have more experience under our belts.

SaraMcCutcheon commented on tests/active_features/env-clone.feature:
```diff
@@ -0,0 +1,20 @@
+
+  @vcr site_clone-content
+  Scenario: Site Clone Environment
+    When I run "terminus env:clone [[test_site_name]].test [[test_site_name]].dev --yes"
```
This should be
`When I run "terminus env:clone [[test_site_name]].test dev"`

SaraMcCutcheon commented on tests/new_unit_tests/Commands/Env/MockCloneCommand.php:
```diff
@@ -0,0 +1,8 @@
+
+use Pantheon\Terminus\Commands\CloneCommand;
+
+class MockCloneCommand extends CloneCommand
```
Is this going to be put to use in a subsequent PR?

dustinleblanc replied:
This is simply a test class so that I can test the protected methods of the abstract class

SaraMcCutcheon commented on tests/new_unit_tests/Commands/Auth/CloneCommandTest.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        parent::setUp();
+        $this->command = new MockCloneCommand();
+        $this->site = $this->getMockBuilder(Site::class)
```
This is redundant. `CommandTestCase::setUp()` does this exact operation.
https://github.com/pantheon-systems/terminus/blob/master/tests/new_unit_tests/Commands/CommandTestCase.php#L258-L260

dustinleblanc replied:
Sweet! didn't realize that, yay!

dustinleblanc commented:
@TeslaDethray thanks for the feedback! I'll take another pass as soon as I have time, hopefully today. Let me know if you guys get blocked waiting for me, I am more than happy to pass off to someone else if we need to :)

SaraMcCutcheon commented on src/Commands/CloneCommand.php:
```diff
@@ -0,0 +1,191 @@
+
+    const PROGRESS_FORMAT = " <bg=blue;fg=black> %message% </>\n <info>Progress: %bar% <info>%percent:3s%%</info> \n <info>Elapsed: %elapsed:6s% \n Estimated Time Remaining: %estimated:-6s%</info>";
+    protected $operations = ['cloneFiles', 'cloneDatabase'];
+    protected $originEnv;
```
To keep things standardized, all vars/properites should be done in underscore_case and function names in camelCase.

SaraMcCutcheon reviewed (changes requested):
Other than the detail about underscore_case and camelCase, this looks good to me. I am highly enamored of your progress bar! After this is merged, would you care to extract its logic into a separate class we can make use of?

dustinleblanc commented:
@TeslaDethray vars have been updated

dustinleblanc commented:
@TeslaDethray also, yeah, I was curious about how we want to abstract the progress bar. @greg-1-anderson if you have any ideas based on Robo work I'd love to get your opinion too. I'd love to maybe work with @kimby77 at some point on defining the 'ui' so to speak for all of our style components in a styleguide that could just be called from all commands. Symfony is just so awesome at making the output look killer.

kimby77 commented:
@dustinleblanc  oh do you mean this??? 

https://docs.google.com/document/d/11YeUoVbnGH01SyQt2TT-TwxrmXs91n0drMEpqDQnF7E/edit

dustinleblanc commented:
@kimby77 Yeah...I'd love to help take the standards you write up and turn them into PHP code that defines the canonical style so all the other developers can just use it. Granted, I have to work around the 'day job' so to speak, so the amount of time I get to commit to Terminus / development is lacking, but I'd just love to assist, I find it to be a lot of fun. The progress bar that I used in this PR is just delightfully fun...but I have zero idea if it fits our styleguide 🎱 . It looks like 
![image](https://cloud.githubusercontent.com/assets/3948808/19082231/43320308-8a2b-11e6-8898-5b5593c5b216.png)

kimby77 commented:
👍 @dustinleblanc!  We'll need to coordinate with Sarah to prioritize work for this style guide.  It has been most useful for us so far in helping us define the standards that we want for the official 1.0 version.  But this still needs some work, and preparation to be made public alongside the 1.0 stable release... as Terminus is open source, as both pantheors and outside devs contribute code to terminus, this will help maintain consistency so that it remains reliable and trustworthy for our users.

greg-1-anderson commented:
If you use Robo tasks, then progress bars are handled for you automatically.
- The progress bar does not start until a specified amount of time has elapsed (default: 2s), so very short operations do not display a progress bar.
- If you have multiple tasks running in a collection, the progress bar will span the duration of all of the tasks (one longer-duration bar, rather than N bars that go to 100% for each operation).
- If you print log messages in the middle of a progress operation, then the progress bar will be hidden and re-shown again after the log message is displayed, so the progress bar continues to display at the bottom of the page, as messages appear above it.

If you do not use Robo tasks, then you can do progress bars as shown in this PR.  This is more appropriate if you are not going to have a lot of progress bars, & do not want to implement everything as a Robo task.

greg-1-anderson commented:
@dustinleblanc @kimby77 Speaking of the style guide, Terminus is currently using Robo styles for log messages. This can be customized if desired.

dustinleblanc commented:
Yep, The robo styles are actually where the progressbar itself is coming from, I just added some colors, split it into multi line, and added some tokens to be output. Just chatted with Ben about potential a potential home for all the 'stylin' and I'm still fairly open to where it lives in the end

greg-1-anderson commented:
@dustinleblanc $this->io() is just the SymfonyStyle object.

dustinleblanc commented:
@greg-1-anderson cool, so if I create a new Symfony style object we can inject that into all commands. sa-weet :)

greg-1-anderson commented:
@dustinleblanc Hm, the SymfonyStyle object is just constructed from the input and output objects in the IO trait; it's not managed by DI right now.  But if you want to subclass SymfonyStyle, we can change the way this works.

greg-1-anderson commented:
If Terminus isn't going to use Robo tasks, then maybe we should separate off the parts it is using.
